academy/smart-lock-with-arduino-part2.md

```diff
@@ -0,0 +1,480 @@
+
+## Modify the Sketch
+
+The first thing to do is to update the sketch to toggle the status of the lock in the smart contract. Let's look at the `SmartLockDevice.ino` in the `SmartLockDevice` directory. 
+
+The first thing to do here is to add a global variable to help us store the previous status of the lock, as well defining the button pin. 
+
+Add this code after the execution action: 
+
+```cpp
+// Global variable to store previous status for toggling the lock
+static bool previousStatus = false;
+
+// The button pin
+#if defined(__SAMD21G18A__)
+#define BUTTON_PIN 3    // Pin D3
+#else
+#define BUTTON_PIN 18
+#endif
+```
+At this point, we need to create a function to toggle the status of the lock in the smart contract: 
+
+```cpp
+// Toggles the status of the lock in the smart contract
+void toggleStatusOnBlockchain()
+{
+    Serial.println("Toggling lock status");
+    // Convert the privte key to a byte array
+    const char pK[] = SECRET_PRIVATE_KEY;
+    uint8_t pk[IOTEX_PRIVATE_KEY_SIZE];
+    signer.str2hex(pK, pk, IOTEX_PRIVATE_KEY_SIZE);
+
+    // Create the account and get the nonce
+    Account originAccount(pk);
+    AccountMeta accMeta;
+    ResultCode result = connection.api.wallets.getAccount(fromAddress, accMeta);
+    if (result != ResultCode::SUCCESS)
+    {
+        Serial.print("Error getting account meta: ");
+        Serial.print(IotexHelpers.GetResultString(result));
+    }
+    int nonce = atoi(accMeta.pendingNonce.c_str());
+
+    // Construct the action - Create the parameters
+    ParameterValue paramOpen;
+        paramOpen.value.boolean = !previousStatus;
+        paramOpen.type = EthereumTypeName::BOOL;
+    ParameterValuesDictionary params;
+    params.AddParameter("open", paramOpen);
+
+    // Contruct the action - Generate contract call data
+    String callData = "";
+    contract.generateCallData("setState", params, callData);
+
+    // Send the action and store it's hash for printing it to the console
+    uint8_t hash[IOTEX_HASH_SIZE] = {0};
+    result = originAccount.sendExecutionAction(connection, nonce, 20000000, "1000000000000", "0", contractAddress, callData, hash);
+
+    // If successful print the action has, otherwise print an error message
+    if (result == ResultCode::SUCCESS)
+    {
+        Serial.print("Hash: ");
+        for (int i=0; i<IOTEX_HASH_SIZE; i++)
+        {
+            char buf[3] = "";
+            sprintf(buf, "%02x", hash[i]);
+            Serial.print(buf);
+        }
+        Serial.println();
+    }
+    else
+    {
+        Serial.println("Failed to toggle lock status");
+    }
+}
+```
+
+Now add the following bit of code to keep track of when the button is pressed, and to interrupt the service runtime accordingly:
```

Comment:
In case the end user is not familiar with how interrupts work. It might be worth describing this with a bit more detail. Something like:

We will be using an interrupt to detect button press. The following line enables the interrupt on the button pin and configures the `isr()` function as the interrupt service routine:
```cpp
attachInterrupt(BUTTON_PIN, isr, FALLING);
```
When the button is pressed, the interrupt service routine will be executed.
It is a good practice to do as little work as possible from within the isr in order to not hang the main execution process. In our case, we simply set a flag `buttonPressed` and return to the main execution process, which will take care of checking this flag and acting accordingly. Here's the ISR that we will be using:
```cpp
// Flag that is set on button press
volatile bool buttonPressed = false;
// Interrupt service routine that is triggered when the button is pressed 
#if defined(ESP32)
void IRAM_ATTR isr() {
#else
void isr() {
#endif
    detachInterrupt(BUTTON_PIN);
        buttonPressed = true;
    attachInterrupt(BUTTON_PIN, isr, FALLING);
}
```